File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        // If p and q > curr, go right
        if (p->val >= root->val && 
            q->val >= root->val) {
            if (root->val == q->val || root->val == p->val)
            {
                return root;
            }
            if (root->right != nullptr) {
                return lowestCommonAncestor(root->right, p, q);
            }
        }
        // if p and q < curr, go left
        if (p->val <= root->val && 
            q->val <= root->val) {
            cout << "hello";
            if (root->val == q->val || root->val == p->val)
            {
                return root;
            }
            if (root->left != nullptr) {
                return lowestCommonAncestor(root->left, p, q);
            }
        }
        cout << "hello";
        return root;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        self.goodcount = 0
        self.goodDFS(root, float('-inf'))
        return self.goodcount
        
    def goodDFS(self, root: TreeNode, max):
        if not root: return
        if root.val >= max:
            max = root.val
            self.goodcount += 1
        self.goodDFS(root.left, max)
        self.goodDFS(root.right, max)